src/api/libcellml/componententity.h

```diff
@@ -0,0 +1,189 @@
+     *
+     * This is an overloaded method.
+     * @param index The index of the Component to return (zero-based).
+     * @return A reference to the Component at the given index.
+     */
+    Component_Ptr getComponent(size_t index);
+
+    /**
+     * @brief Get a component with the given name @p name.
+     * Returns a const reference to a component with the name @p name.  If the
+     * name is not valid a std::out_of_range exception is thrown.
+     *
+     * This is an overloaded method.
+     * @param name The name of the Component to return.
+     * @return A const reference to the Component with the given name.
+     */
+    const Component_Ptr& getComponent(const std::string &name) const;
+
+    /**
+     * @brief Get a component with the given name @p name.
+     * Returns a reference to a component with the name @p name.  If the name
+     * is not valid a std::out_of_range exception is thrown.
+     *
+     * This is an overloaded method.
+     * @param name The name of the Component to return.
+     * @return A reference to the Component with the given name.
+     */
+    Component_Ptr getComponent(const std::string &name);
+
+    /**
+     * @brief Take the component at the given index and return it.
+     * Removes the component at the given index position and returns it.
+     * If an invalid index is passed to the method a std::out_of_range
+     * exception is thrown.
+     * @param index The index of the Component to take (zero-based).
+     * @return The Component at the given index.
+     */
+    Component_Ptr takeComponent(size_t index);
+
+    /**
+     * @brief Take the component with the given name @p name and return it.
+     * Takes the component with the given name @p name and returns it.
+     * If an invalid name is passed to the method a std::out_of_range
+     * exception is thrown.
+     *
+     * This is an overloaded method.
+     * @param name The name of the Component to take.
+     * @return The Component identified with the given name.
+     */
+    Component_Ptr takeComponent(const std::string &name);
+
+    /**
+     * @brief Replace a component at index.
+     * Replaces the component at index @p index with @p c.  Index must be a valid index
+     * position.
+     * @param index Index of the Component to replace (zero-based).
+     * @param c The Component to use for replacement.
+     */
+    void replaceComponent(size_t index, const Component_Ptr &c);
+
+    /**
+     * @brief Replace a component with the given name.
+     * Replaces the component with the given name @p name with @p c.  Name must be a valid name
+     * of a component in the Component, if it isn't a std::out_of_range exception is thrown.
+     *
+     * This is an overloaded method.
+     * @param name The name of the Component to replace.
+     * @param c The Component to use for replacement.
+     */
+    void replaceComponent(const std::string &name, const Component_Ptr& c);
```

Comment:
```void replaceComponent(const std::string &name, const Component_Ptr &c);```

rather than

```void replaceComponent(const std::string &name, const Component_Ptr& c);```